naga/src/front/wgsl/parse/mod.rs

```diff
@@ -2433,4 +2454,43 @@ impl Parser {
         }
         Ok(brace_nesting_level + 1)
     }
+
+    fn diagnostic_control<'a>(&self, lexer: &mut Lexer<'a>) -> Result<(), Error<'a>> {
+        // TODO(maybe): self.push_rule_span(Rule::???, lexer)
+        // TODO: I see `*_argument` helpers, but they're not used here. Guessing it's
+        // more valuable to make something easy to validate against the WGSL spec.?
+        lexer.expect(Token::Paren('('))?;
+        let (severity_control_name, severity_control_name_span) = lexer.next_ident_with_span()?;
+        match severity_control_name {
+            "error" | "warning" | "info" | "off" => (/* neat */),
+            _ => todo!(
+                "diagnostic for invalid severity control name {:?}",
+                severity_control_name
+            ),
+        }
+        lexer.expect(Token::Separator(','))?;
+        let (diagnostic_rule_name, diagnostic_rule_name_span) = lexer.capture_span(|lexer| {
+            // TODO: ensure that errors are good
+            let mut segments = Vec::new();
+            let (first_segment, _span) = lexer.next_ident_with_span()?;
+            segments.push(first_segment);
+            while lexer.skip(Token::Separator('.')) {
+                let (segment, _span) = lexer.next_ident_with_span()?;
+                segments.push(segment);
+            }
```

Comment:
There are, at most, two segments here (called "diagnostic name tokens" upstream, maybe rename?), per [spec.](https://www.w3.org/TR/WGSL/#diagnostic-triggering-rule):

> The name of a [triggering rule](https://www.w3.org/TR/WGSL/#diagnostic-triggering-rule) is either:
> 
> -   a [diagnostic\_name\_token](https://www.w3.org/TR/WGSL/#syntax-[diagnostic_name_token](https://www.w3.org/TR/WGSL/#syntax-diagnostic_name_token)), or
>     
> -   two [diagnostic\_name\_token](https://www.w3.org/TR/WGSL/#syntax-[diagnostic_name_token](https://www.w3.org/TR/WGSL/#syntax-diagnostic_name_token)) tokens, separated by a period `'.'` (U+002E).

Reply:
In WGSL, the idea is:

- If you see a single-component diagnostic rule name, it's an error if the name isn't recognized.
- If you see a double-component diagnostic rule name, it's *ignored* if the name isn't recognized. This lets you write `clippy.my_picky_lint` and have browsers ignore it.

So ultimately joining them into a `String` might not be what we want. But there's no need to fix that at this stage.

Reply:
Invalidated by my latest force-push, which does the extract of a single diagnostic name token structurally instead.